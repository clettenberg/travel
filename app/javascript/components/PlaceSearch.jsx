import _ from 'lodash';
import React from 'react';

import StandaloneSearchBox from 'react-google-maps/lib/components/places/StandaloneSearchBox';

class PlaceSearch extends React.Component {
  componentWillMount() {
    this.setState({
      places: [],
    });

    this.csrfToken = document.getElementsByTagName('meta')['csrf-token'].content;
  }

  onKeyPress = (event) => {
    if (event.which == 13) {
      event.preventDefault();
    }
  }

  onPlacesChanged = () => {
    const places = this.searchBox.getPlaces()

    if (_.isEmpty(places)) {
      return;
    }

    this.setState({
      places: _.concat(places, this.state.places),
    });
  }

  render() {
    const SearchBox = props => (
      <span>
        <StandaloneSearchBox
          ref={props.onSearchBoxMounted}
          bounds={props.bounds}
          onPlacesChanged={props.onPlacesChanged}
        >

          <input
            onKeyPress={props.onKeyPress}
            type="text"
            placeholder="Find a Place"
            id="place_search"
            style={{
              boxSizing: 'border-box',
              border: '1px solid transparent',
              width: '240px',
              height: '32px',
              marginTop: '27px',
              padding: '0 12px',
              borderRadius: '3px',
              boxShadow: '0 2px 6px rgba(0, 0, 0, 0.3)',
              fontSize: '14px',
              outline: 'none',
              textOverflow: 'ellipses',
            }}
          />
        </StandaloneSearchBox>

        <ol>
          {props.places.map(({place_id, name}, index) =>
            <span key={index}>
              <input type="hidden" name="trip[place_ids][]" value={place_id}></input>
              <p>{name}</p>
            </span>
          )}
        </ol>
      </span>
    );

    return (
      <SearchBox
        onPlacesChanged={this.onPlacesChanged}
        onSearchBoxMounted={(searchBox) => { this.searchBox = searchBox; }}
        places={this.state.places}
        csrfToken={this.csrfToken}
        onKeyPress={this.onKeyPress}
      />
    );
  }
}

export default PlaceSearch;
