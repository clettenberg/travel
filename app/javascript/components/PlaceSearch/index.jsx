import React from 'react'
import PropTypes from 'prop-types'
import MapquestSearchClient from '../../api/MapquestSearchClient'
import PlaceSearchResults from '../PlaceSearchResults'

const propTypes = {
  onPlaceSelection: PropTypes.func.isRequired
}

class PlaceSearch extends React.Component {
  constructor (props) {
    super(props)

    this.handleNameSearchFormSubmit = this.handleNameSearchFormSubmit.bind(this)
    this.handleLatLonSearchFormSubmit = this.handleLatLonSearchFormSubmit.bind(this)
    this.handleSearchValueChange = this.handleSearchValueChange.bind(this)
    this.handlePlaceSelection = this.handlePlaceSelection.bind(this)
    this.handleSubmit = this.handleSubmit.bind(this)

    this.state = {
      searchType: 'name',
      name: '',
      lat: '',
      lon: '',
      selectedPlaceId: null,
      errors: [],
      placeSearchResults: []
    }
  }

  handleSearchTypeChange (searchType) {
    this.setState({ searchType })
  }

  handleSearchValueChange (event) {
    const { name } = event.target
    this.setState({ [name]: event.target.value })
  }

  handleSubmit (event) {
    event.preventDefault()

    if (this.state.searchType === 'name') {
      this.handleNameSearchFormSubmit()
    } else {
      this.handleLatLonSearchFormSubmit()
    }
  }

  handleNameSearchFormSubmit () {
    MapquestSearchClient
      .findPlacesByName(this.state.name)
      .then((placeSearchResults) => {
        this.setState({ placeSearchResults, errors: [] })
      })
      .catch((error) => {
        this.setState({ placeSearchResults: [], errors: [error] })
      })
  }

  handleLatLonSearchFormSubmit () {
    MapquestSearchClient
      .findPlaceByLatLon({ lat: this.state.lat, lon: this.state.lon })
      .then((placeSearchResults) => {
        this.setState({ placeSearchResults: [placeSearchResults], errors: [] })
      })
      .catch((error) => {
        this.setState({ placeSearchResults: [], errors: [error] })
      })
  }

  handlePlaceSelection (changeEvent) {
    const selectedPlaceId = changeEvent.target.value
    this.setState({ selectedPlaceId })
    const selectedPlace = this.state.placeSearchResults.find(result => (
      result.place_id === selectedPlaceId
    ))
    this.props.onPlaceSelection(selectedPlace)
  }

  render () {
    let placeSearchInput

    const errors = this.state.errors.map(err => (
      <p key={err.message}>{err.message}</p>
    ))

    if (this.state.searchType === 'name') {
      placeSearchInput = (
        <input
          placeholder='Place Name'
          type='text'
          className='form-control'
          name='name'
          value={this.state.name}
          onChange={this.handleSearchValueChange}
        />
      )
    } else {
      placeSearchInput = (
        <React.Fragment>
          <input
            placeholder='Latitude'
            type='text'
            className='form-control'
            name='lat'
            value={this.state.lat}
            onChange={this.handleSearchValueChange}
          />
          <input
            placeholder='Longitude'
            type='text'
            className='form-control'
            name='lon'
            value={this.state.lon}
            onChange={this.handleSearchValueChange}
          />
          <input
            type='submit'
            style={{ display: 'none' }}
          />
        </React.Fragment>
      )
    }

    return (
      <div>
        <form onSubmit={this.handleSubmit}>
          <div className='input-group mb-3'>
            <div className='input-group-prepend'>
              <button
                onClick={() => this.handleSearchTypeChange('name')}
                className={`btn btn-outline-secondary ${this.state.searchType === 'name' ? 'active' : ''}`}
                id='name'
                type='button'
              >
                Place Name
              </button>
              <button
                onClick={() => this.handleSearchTypeChange('latlon')}
                className={`btn btn-outline-secondary ${this.state.searchType === 'latlon' ? 'active' : ''}`}
                id='latlon'
                type='button'
              >
                Lat/Lon
              </button>
            </div>
            {placeSearchInput}
          </div>
        </form>
        <div className='row'>
          <ul className='col'>
            <PlaceSearchResults
              results={this.state.placeSearchResults}
              selectedId={this.state.selectedPlaceId}
              onPlaceSelection={this.handlePlaceSelection}
            />
          </ul>
          {errors}
        </div>
      </div>
    )
  }
}

PlaceSearch.propTypes = propTypes
export default PlaceSearch
