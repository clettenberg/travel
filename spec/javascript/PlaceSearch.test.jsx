import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import PlaceSearch from '../../app/javascript/components/PlaceSearch';

describe('PlaceSearch Component', () => {
  describe('Place Name', () => {
    it('should render correctly', () => {
      const wrapper = shallow(<PlaceSearch onPlaceSelection={jest.fn()} />);

      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });

  describe('Lat/Lon', () => {
    it('should render correctly', () => {
      const wrapper = shallow(<PlaceSearch onPlaceSelection={jest.fn()} />);
      wrapper.find('button#latlon').simulate('click');

      expect(toJson(wrapper)).toMatchSnapshot();
    });
  });
});
