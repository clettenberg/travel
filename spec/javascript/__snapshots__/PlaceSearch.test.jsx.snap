// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlaceSearch Component Lat/Lon should render correctly 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <div
      className="input-group mb-3"
    >
      <div
        className="input-group-prepend"
      >
        <button
          className="btn btn-outline-secondary "
          id="name"
          onClick={[Function]}
          type="button"
        >
          Place Name
        </button>
        <button
          className="btn btn-outline-secondary active"
          id="latlon"
          onClick={[Function]}
          type="button"
        >
          Lat/Lon
        </button>
      </div>
      <input
        className="form-control"
        name="lat"
        onChange={[Function]}
        placeholder="Latitude"
        type="text"
        value=""
      />
      <input
        className="form-control"
        name="lon"
        onChange={[Function]}
        placeholder="Longitude"
        type="text"
        value=""
      />
      <input
        style={
          Object {
            "display": "none",
          }
        }
        type="submit"
      />
    </div>
  </form>
  <div
    className="row"
  >
    <ul
      className="col"
    >
      <PlaceSearchResults
        onPlaceSelection={[Function]}
        results={Array []}
        selectedId={null}
      />
    </ul>
  </div>
</div>
`;

exports[`PlaceSearch Component Place Name should render correctly 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <div
      className="input-group mb-3"
    >
      <div
        className="input-group-prepend"
      >
        <button
          className="btn btn-outline-secondary active"
          id="name"
          onClick={[Function]}
          type="button"
        >
          Place Name
        </button>
        <button
          className="btn btn-outline-secondary "
          id="latlon"
          onClick={[Function]}
          type="button"
        >
          Lat/Lon
        </button>
      </div>
      <input
        className="form-control"
        name="name"
        onChange={[Function]}
        placeholder="Place Name"
        type="text"
        value=""
      />
    </div>
  </form>
  <div
    className="row"
  >
    <ul
      className="col"
    >
      <PlaceSearchResults
        onPlaceSelection={[Function]}
        results={Array []}
        selectedId={null}
      />
    </ul>
  </div>
</div>
`;
